evap/staff/views.py

```diff
@@ -1461,11 +1449,24 @@ def evaluation_person_management(request, evaluation_id):
                 delete_import_file(request.user.id, import_type)
             elif "copy" in operation:
                 copy_form.evaluation_selection_required = True
-                if copy_form.is_valid():
-                    import_evaluation = copy_form.cleaned_data["evaluation"]
-                    importer_log = import_persons_from_evaluation(
-                        import_type, evaluation, test_run=False, source_evaluation=import_evaluation
+                if not copy_form.is_valid():
+                    return render(
+                        request,
+                        "staff_evaluation_person_management.html",
+                        {
+                            "semester": evaluation.course.semester,
+                            "evaluation": evaluation,
+                            "participant_excel_form": participant_excel_form,
+                            "participant_copy_form": participant_copy_form,
+                            "contributor_excel_form": contributor_excel_form,
+                            "contributor_copy_form": contributor_copy_form,
+                            "importer_log": importer_log,
+                        },
```

Comment:
this is a bit duplicated but I don't know of any other way to redirect to the current page with the invalid copy form for the error. Any ideas?

Reply:
you could only write it once by defining
```python3
render_self = lambda: render(
    ...
)
```
but I don't know if that is an improvement

Reply:
The standard django control flow would be to just fall through to the default `render` call at the end of the method in case any form fails to validate.

The control flow of the view was designed with the idea that a non-test operation cannot fail, so a fallthrough from the import operation would always end in the `redirect` that currently prevents us from applying the above logic here. Maybe, now that it is more complex, it should be split up into more components anyway, or maybe the redirect should be moved, idk.

We could accept duplicating the call and just try to deduplicate the kwargs. However, the `test_passed` args depend on the test run happening below.

Random thought: An immediately invoked function could use `return` to "jump" to the end while skipping the offending redirect -- I've never seen that done in python though, I think the syntax would require giving the function a name anyway?

Reply:
Yeaaaaah, I think immediately invoked function is not really a thing I think and I also think that we would have to give a name

Green light from me for a bigger refactor :D

Reply:
I extracted a `import_or_copy_participants` returning the success of the operation. If successful, we can easily redirect to the semester page, otherwise just fallthrough to the common render() call. What do you think about it?

Reply:
We could inline `copy_form = participant_copy_form if "participants" in operation else contributor_copy_form` or move it to the function as well. We wouldn't need the pylint ignore then, but I don't think this is better...

Reply:
I think moving into the function is not the right move, I think leaving it as is or inlining to the call (maybe together with https://github.com/e-valuation/EvaP/pull/2198#discussion_r1641942557) both sounds fine to me :)